notebooks/03-query/15-pivot-unpivot.ipynb

```diff
@@ -0,0 +1,372 @@
+    "\n",
+    "The result is much cleaner than the prior query and it is easy to determine which channels have the most robotic activity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "65c489a8-a7f2-4cc0-b30c-f30d8ed32be2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT \"channel\", \n",
+    "        SUM(\"robot_added\")       AS \"added_robot\", \n",
+    "        SUM(\"robot_deleted\")     AS \"deleted_robot\", \n",
+    "        SUM(\"human_added\")       AS \"added_human\",\n",
+    "        SUM(\"human_deleted\")     AS \"deleted_human\", \n",
+    "        \n",
+    "        SAFE_DIVIDE( SUM(\"human_deleted\" + \"human_added\") * 1.0 , \n",
+    "                     SUM( \"robot_deleted\" + \"robot_added\" + \"human_deleted\" + \"human_added\" )\n",
+    "                     \n",
+    "                    )            AS \"human_ratio\"\n",
+    "                    \n",
+    "FROM\n",
+    "(\n",
+    "    SELECT\n",
+    "      \"channel\", \n",
+    "       COALESCE(\"robot_added\",0) AS \"robot_added\", \n",
+    "       COALESCE(\"human_added\",0) AS \"human_added\", \n",
+    "       COALESCE(\"robot_deleted\",0) AS \"robot_deleted\", \n",
+    "       COALESCE(\"human_deleted\",0) AS \"human_deleted\"\n",
+    "    FROM \"example-pivot-unpivot\" \n",
+    "    \n",
+    "    PIVOT ( SUM(added) as added, \n",
+    "            SUM(deleted) as deleted \n",
+    "            \n",
+    "            FOR \"isRobot\" IN ('true' as robot, 'false' as human)\n",
+    "          )\n",
+    ")x\n",
+    "GROUP BY 1\n",
+    "ORDER BY 6\n",
+    "'''\n",
+    "display_client.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3f0af833-9508-4f26-a484-2778ca857a2a",
+   "metadata": {},
+   "source": [
+    "## UNPIVOT example\n",
+    "Given the prior query we can now investigate the sources of the updates in the most robotically updated pages.\n",
+    "```\n",
+    "  UNPIVOT ( \"changes\"  FOR \"action\" IN (\"added\", \"deleted\") )\n",
+    "```\n",
+    "This example uses UNPIVOT to:\n",
+    "- take the values of multiple columns: `\"added\",\"deleted\"`\n",
+    "- incorporate them into a single column called the \"values column\": `\"changes\"`\n",
+    "- while expanding the result with a \"names column\" containing the names of the pivoted columns: `\"action\"`\n",
+    "\n",
+    "The columns names become row values.\n",
+    "\n",
+    "In this example we examine the most active user as measured by total adding or deleting of text in the wikipedia pages in the channel `'#it.wikipedia'` which had a ratio of about 0.50 (half robots):"
```

Comment:
Make active:

"Run the following cell to use UNPIVOT to sum the total changes by user".